### PR TITLE
Tests log output improvements 

### PR DESCRIPTION
**Pull request by dimitrovmaksim**
Issues:
closes https://github.com/LimeChain/matchstick/issues/123
closes https://github.com/LimeChain/matchstick/issues/145

- [x] Display all failed tests at the end of the log output, before the summary.
- [x]  Move the error message and the backlog there.
- [x]  Adds a timestamp at the end of the log before `Program executed in`
- [x]  Adds the time it took for each test to execute on the same line
~~- [x] Display `path/to/file:line` for each failed test~~ - Will be addressed in a separate Issue/PR

Screenshot:
Screenshot 2021-12-03 at 11 04 41

### CODE REVIEW

File: src/test_suite.rs [hunk 1]
```diff
@@ -67,13 +69,19 @@ impl Test {
             }
             Box::new([wasmtime::Val::I32(0)])
         });
+        let elapsed = now.elapsed();
+        // Convert the elapsed time to milliseconds
+        // Seems hacky, might need refactoring
+        let elapsed_in_ms = elapsed.as_secs_f32() * 1000.0;
```
Comment (georg-getz) [inline src/test_suite.rs:75]: elapsed.as_millis();

Comment (dimitrovmaksim) [reply src/test_suite.rs:75]: `as_millis()` yields an int, and most of the tests complete for less than 1ms and the results is 0ms. Only `as_secs` has a f32/f64 option.

Comment (dimitrovmaksim) [reply src/test_suite.rs:75]: Another way that looks less hacky
```
let millis = elapsed.as_millis();
let micros = elapsed.as_micros();
format!("{} - {:?}.{:?}ms", self.name.clone(), millis, micros);
```

Comment (VIVelev) [reply src/test_suite.rs:76]: Maybe you can move `1000.0` at the top as a const, something like `SECS_TO_MILLIS`, but I don't think it's a big deal.
You can also join the above two rows.

File: src/test_suite.rs [hunk 2]
```diff
@@ -67,22 +70,25 @@ impl Test {
...
+
         logging::sub_indent();
         let logs = logging::flush();
 
+        let msg = format!("{} - {:.3?}ms", self.name.clone(), elapsed_in_ms);
```
Comment (VIVelev) [inline src/test_suite.rs:81]: It would be cool to print `elapsed_in_ms` in a different color, so it's more noticeable in the logs. :)

**Review by VIVelev — CHANGES_REQUESTED** [review]
Nice PR!

File: src/main.rs [hunk 3]
```diff
@@ -241,15 +242,21 @@ ___  ___      _       _         _   _      _
...
         for test in &val.tests {
-            if test.run().passed {
+            let result = test.run();
+            if result.passed {
                 passed_tests += 1;
             } else {
+                all_failed_tests.insert(test.name.clone(), result.logs);
                 failed_tests += 1;
             }
         }
```
[the remark this answers is not in the document]
Comment (dimitrovmaksim) [reply src/main.rs:262]: I was working on displaying the full path to the `test()` function (as in the screenshot), and one of the issues was that there might be some generic test names (e.g. "Should yield correct value") in different suites, so this might be the right way to do it. I will look into it today. Thanks!

Screenshot 2021-12-02 at 9 33 23

File: src/test_suite.rs (hunk 1 above)
Comment (dimitrovmaksim) [reply src/test_suite.rs:76]: Maybe we could extract the logic into a trait for the std::time::Duration struct and add a method `as_millis_f64()`

File: src/tests_location.rs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+                .into_iter()
+                .filter_map(|line| {
+                    line_number += 1;
+                    let line_as_string = line.as_ref().unwrap();
+
+                    if line.as_ref().unwrap().starts_with("test(") {
+                        // This regex finds the beginning of the test name
+                        // It looks for (", but also handles situations if there are
+                        // white spaces between the opening ( and "
+                        let start = Regex::new(r#"\(\s*""#)
```
Comment (VIVelev) [inline src/tests_location.rs:56]: What if I have:
```ts
test("Can call mappings (\"the source code\") with custom events", () => {
  ...
  assert.fieldEquals(GRAVATAR_ENTITY_TYPE, "0xdead", "id", "0xdedd")
  ...
}
```

I personally feel like regex are susceptible to introducing bugs, if you can somehow find a solution that does not relay on regex would be great!

A promising direction may be [AS debugging](https://www.assemblyscript.org/debugging.html), although, as we know, AS is not that mature, so idk.

What do you think?

File: src/tests_location.rs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+                            .unwrap()
+                            .find(line_as_string)
+                            .unwrap()
```
Comment (VIVelev) [inline src/tests_location.rs:59]: I know it's annoying, but try using `.expect(...)` or `.unwrap_or_else(|err| ...)` instead of `.unwrap()`, since often times `.unwrap()` does not provide enough info for debugging.
However, there are some cases, like the `.unwrap()` above, where it's totally fine to use `.unwrap()`, since there is no way it will fail, we can obviously see that the regex expr is a valid one.

**Review by VIVelev — CHANGES_REQUESTED** [review]

File: src/tests_location.rs (hunk 4 above)
Comment (VIVelev) [reply src/tests_location.rs:56]: Oh, wait... I think a solution is to take the first `("` match and the last `",`

Or actually no... What if you have:
```ts
test(
  "\
    Multi-line string:\
    Can call mappings (\"source\") with custom events\
  ", () => {

  ...
}
```

I suppose you see where this is going. Code formatting is fuzzy, which makes regex operating over code hard to write and thus bug-prone.

Another direction I see is to write a mini parser to extract the name.

File: src/tests_location.rs [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            let test_suite = v.get(2).unwrap().replace(".test.ts", "");
+
+            let test_paths: HashMap<String, String> = reader
+                .lines()
+                .into_iter()
+                .filter_map(|line| {
+                    line_number += 1;
+                    let line_as_string = line.as_ref().unwrap();
+
+                    if line.as_ref().unwrap().starts_with("test(") {
```
Comment (VIVelev) [inline src/tests_location.rs:52]: What if you have:
```ts
function testGroup() {
  test(...)
  test(...)
  test(...)
}

testGroup()
```

Also you could have:
```ts
import test as myTest from "..."
```

Maybe even more edge cases...

It turns out to be more complicated than it seems.
If it's possible to do something through [AS debugging](https://www.assemblyscript.org/debugging.html) all of this could be avoided.
We can talk if you want :D

**Review by VIVelev — CHANGES_REQUESTED** [review]
A few more, sorry 😄

File: src/tests_location.rs (hunk 4 above)
Comment (dimitrovmaksim) [reply src/tests_location.rs:56]: The regex catches `test("Can call mappings (\"the source code\") with custom events", () => {` correctly, but the escapes are kept, which is a problem. The multilines would be a problem, but for other reasons, the whole logic so far does not expect a multiline and assumes it's one.

File: src/tests_location.rs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+            files.push(path);
+        } else if path.is_dir() {
+            let mut sub_files: Vec<std::path::PathBuf> =
+                collect_files(&path.to_str().unwrap().to_string());
+            files.append(&mut sub_files);
+        }
+    }
+
+    files
+}
```
[the remark this answers is not in the document]
Comment (dimitrovmaksim) [reply src/tests_location.rs:25]: I'll look into it

File: src/tests_location.rs (hunk 6 above)
Comment (dimitrovmaksim) [reply src/tests_location.rs:52]: I don't know if it's not better to leave this out for a moment, specifically the part to accumulate and display the path and line of the tests and come back to it later with a generally better solution.

Comment (VIVelev) [reply src/tests_location.rs:52]: I think it's a good idea. You can discuss it with the rest of the team on the daily today. Then, go ahead and open an issue for it!

I will now approve this PR, since the rest is ready to go.

**Review by VIVelev — APPROVED** [review]

File: src/tests_location.rs (hunk 4 above)
Comment (dimitrovmaksim) [reply src/tests_location.rs:56]: > Oh, wait... I think a solution is to take the first (" match and the last ",

This is what the regexes look for, but also handle white spaces between the opening `(` and `"` and between the closing `"` and `,`. Still the other issues remain.

**Review by georg-getz — APPROVED** [review]
